pkg/cmd/run/download/download.go

```diff
@@ -0,0 +1,129 @@
+package download
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+
+	"github.com/MakeNowJust/heredoc"
+	"github.com/cli/cli/pkg/cmdutil"
+	"github.com/cli/cli/pkg/iostreams"
+	"github.com/spf13/cobra"
+)
+
+type DownloadOptions struct {
+	IO       *iostreams.IOStreams
+	Platform platform
+
+	RunID          string
+	DestinationDir string
+	FilePatterns   []string
+}
+
+type platform interface {
+	List(runID string) ([]Artifact, error)
+	Download(url string, dir string) error
+}
+
+func NewCmdDownload(f *cmdutil.Factory, runF func(*DownloadOptions) error) *cobra.Command {
+	opts := &DownloadOptions{
+		IO: f.IOStreams,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "download [<run-id>]",
+		Short: "Download artifacts generated by a workflow run",
+		Args:  cobra.MaximumNArgs(1),
+		Example: heredoc.Doc(`
+		  # Download all artifacts generated by a workflow run
+		  $ gh run download <run-id>
```

Comment:
all of the other `gh run` subcommands drop into the interactive gflow without arguments. if this is run with no arguments (ie no `--pattern` and no run id), do you think it's valid to drop into that prompt here, too?

Reply:
Yeah, I was wondering about what could the interactive flow be about! Do you think it should:
1. Display the list of recent runs;
2. Display the list of available per-repository artifacts?

Reply:
good question, sorry I missed you asking this in the PR body :facepalm: My thinking is something like:

```
$ gh run download
? Would you like to download artifacts for a specific run?
> No, let me choose repository level artifacts
   Run 1
   Another Run
? Which artifact would you like to download?
> Artifact 1
   Artifact 2
```

Reply:
Thanks! Since first choosing a run and then choosing an artifact _within_ that run would just present a subset of options that would otherwise be present when scoped across the whole repo, I went with just prompting with a multi-select containing last 10 non-expired artifacts in a repository.

Is multi-select even what we want, or do you think people would mostly just be interested in a single artifact in interactive mode? If so, we can change it to single-option select and rely on non-interactive mode for multi-archive download.

Reply:
I like this approach! I think it makes sense and I don't have a strong sense one way or the other which people would prefer so I'm cool with keeping it this way and see what the reception is.